### ISSUE TITLE
Automatically gzip large responses

### ISSUE THREAD

**Issue by jaulz** [brefphp/bref#1661]
I wonder if Bref could automatically try to gzip the response if it's bigger than the 6MB limit. I gave it a try and at least this approach works for me (removed the Vapor header):
https://bannister.me/blog/gzip-compression-on-laravel-vapor/

Since it's already possible via (Laravel) middleware, it's not urgent but maybe a nice addition.

**Comment by mnapoli** [comment]
Hey @jaulz, that is interesting! Do you have any numbers on bigger payloads that you are able to return that way?

**Comment by jaulz** [comment]
Yeah, for the one critical response I was dealing with it's a factor of 24 (from 8MB of text to 300KB):
![image](https://github.com/brefphp/bref/assets/5358638/a08cd1f7-e42f-44e6-b2d6-ea28704159bd)

I used this code:
```php
      $content = $response->getContent();
      $compressedContent = gzencode($content, 9);

      $response->setContent($compressedContent);
      $response->headers->add([
        'Content-Encoding' => 'gzip',
        'X-Original-Size' => mb_strlen($content, '8bit'),
        'X-Compressed-Size' => mb_strlen($compressedContent, '8bit'),
      ]);
```

**Comment by kevincerro** [comment]
Now AWS supports response streaming to bypass 6MB limit
- additional cost and with a soft limit of 20MB (maybe ampliable on request)

Its supported on NodeJS official runtimes but AWS allows to support on custom runtimes too.

Maybe this will be interesting also in combination with gzip encoding, because GZIP works great on text responses but on binary responses maybe not

https://aws.amazon.com/es/blogs/compute/introducing-aws-lambda-response-streaming/
https://docs.aws.amazon.com/lambda/latest/dg/configuration-response-streaming.html

**Comment by kevincerro** [comment]
In my case from 87,6kB to 5,8kB on a simple SF project using GZIP

![Captura de pantalla 2023-10-24 a las 14 11 49](https://github.com/brefphp/bref/assets/13980708/e2bf9c90-b074-4d86-8fc2-172467575e14)

**Comment by mnapoli** [comment]
@kevincerro FYI https://github.com/brefphp/bref/issues/1503

**Comment by victormacko** [comment]
> Yeah, for the one critical response I was dealing with it's a factor of 24 (from 8MB of text to 300KB): ![image](https://github.com/brefphp/bref/assets/5358638/a08cd1f7-e42f-44e6-b2d6-ea28704159bd)
> 
> I used this code:
> 
>       $content = $response->getContent();
>       $compressedContent = gzencode($content, 9);
> 
>       $response->setContent($compressedContent);
>       $response->headers->add([
>         'Content-Encoding' => 'gzip',
>         'X-Original-Size' => mb_strlen($content, '8bit'),
>         'X-Compressed-Size' => mb_strlen($compressedContent, '8bit'),
>       ]);

Cheers for that @jaulz  - I was having "Internal Server Error" messages show up when the uncompressed response was around 5mb, but somehow hitting the Lambda response-size limit of 6mb (i'm guessing after base64 encoding from bref). Compressing using gzip within Symfony for me for responses more than 4.5mb made it all work.

**Comment by mnapoli** [comment]
@victormacko thank you for sharing. Did you notice an impact on the response time/Lambda duration?

**Comment by victormacko** [comment]
@mnapoli nope, was pretty much the same, and i'm only doing it for the larger responses to get around the max-lambda-response limitation without having to implement response-streaming

**Comment by deleugpn** [comment]
I have used a lot gzip for 6MB response and its pretty neat. I'd be in favor of a native Bref solution

**Comment by mnapoli** [comment]
Thx all for the details, I've added this to https://github.com/brefphp/bref/pull/1953 as an idea to consider. That looks simple to implement though need to get that right with the base64 encoding.

Sorry I'll be asking very dumb questions but I don't want to make mistakes:

- returning with gzip means you've enabled base64 responses, right?
- I wonder if bref could always gzip + always base64-encode the response, and this would always be smaller than the non-zipped non-base64-encoded version 🤔 
- does that mean the client needs to support gzip responses? I'm pretty sure the answer is yes and that's the reason we wouldn't always enable this… but I want to make sure I'm not missing anything

**Comment by deleugpn** [comment]
> Sorry I'll be asking very dumb questions but I don't want to make mistakes:

No such thing 😂 

> returning with gzip means you've enabled base64 responses, right?

Right! Without it, it doesn't work

> I wonder if bref could always gzip + always base64-encode the response, and this would always be smaller than the non-zipped non-base64-encoded version 🤔

Technically you could, but that may mean you end up flipping it around by having it on-by-default and maybe there's a 0.0001% use-case that would need to opt-out of it. Considering Bref 3 that allows a little more room for changes, I think it could be a good thing as it's a lot more likely that it would work well for those who need it without having any negative impact for those who don't need it. The only caveat is whether you want to break parity with other runtimes as none of them do this "by default". In general, I'd be in favor as its a nice DX to have.

> does that mean the client needs to support gzip responses? I'm pretty sure the answer is yes and that's the reason we wouldn't always enable this… but I want to make sure I'm not missing anything

Yes, but every browser and major programming language support it and I believe it would be very hard to find a flow that wouldn't just work seamless, but I its possible. One potential solution here is to check for `Accept-Encoding` header which supposedly should have `gzip` if the client supports it.

There's a caveat that with gzip, people may end up asking for different (newer) compressing algorithms which again falls in the realm of "maybe we don't support it, but then we need a way to opt-out of it so that folks can compress using whatever they want".

One minor note is that its _probably_ a good idea to check if the response is already gzipped so that folks upgrading to to Bref 3 wouldn't need to make any changes. But if this proves to be any burden at all, I'd also be in favor of just making an upgrade note that "if you gzip your response, you will need to stop doing that otherwise it will break".


Here is my Middleware implementation:

```
final readonly class GzipResponseMiddleware
{
    public function handle(Request $request, Closure $next)
    {
        /** @var \Illuminate\Http\JsonResponse $response */
        $response = $next($request);

        $accept = $request->header('Accept-Encoding');

        if (! str_contains((string) $accept, 'gzip')) {
            return $response;
        }

        if ($response->headers->has('Content-Encoding')) {
            return $response;
        }

        $content = $response->getContent();

        if (! $content) {
            return $response;
        }

        $response->setContent(gzencode($content, 9));

        $response->headers->set('Content-Encoding', 'gzip');

        $response->headers->set('Content-Length', (string) strlen($content));

        return $response;
    }
}
```

**Comment by mnapoli** [comment]
@deleugpn thanks!

I'm thinking about a very dumb scenario: I deploy a web app or API, and then I `curl` it in my terminal: instead of a JSON response I get a gzip binary response, right?

Same e.g. if I use any PHP HTTP client and hit the API?

Or is this transparently handled?

**Comment by deleugpn** [comment]
> I deploy a web app or API, and then I curl it in my terminal: instead of a JSON response I get a gzip binary response, right?

Depends on the implementation. The implementation I shared above requires the request client to include the header `Accept-Encoding: gzip`. If you specify this, then you get a binary response, otherwise its exactly like a regular request. Browsers always include Accept-Encoding. 

So in a very simple / dumb way, you're naturally not going to specify "Accept-Encoding" and as such you get the non-compressed behavior. Here are a demonstration of it:

<details>

<summary>Regular request (No Accept-Encoding headers, JSON Response)</summary>

```
➜ curl --header "Authorization: Bearer xxx" https://example.com/my-endpoint -vvv > r1.txt 2>&1

```

r1.txt
```
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                              Dload  Upload   Total   Spent    Left  Speed

0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host example.com:443 was resolved.
* IPv6: (none)
* IPv4: xx.xxx.xx.xxx, yy.yy.yy.yy
*   Trying xx.xxx.xx.xxx:443...
* Connected to example.com (xx.xxx.xx.xxx) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
} [351 bytes data]
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* (304) (IN), TLS handshake, Unknown (8):
{ [19 bytes data]
* (304) (IN), TLS handshake, Certificate (11):
{ [3849 bytes data]
* (304) (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* (304) (IN), TLS handshake, Finished (20):
{ [36 bytes data]
* (304) (OUT), TLS handshake, Finished (20):
} [36 bytes data]
* SSL connection using TLSv1.3 / AEAD-AES128-GCM-SHA256 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.execute-api.us-west-2.amazonaws.com          <----
*  start date: May 20 00:00:00 2025 GMT
*  expire date: Jun 16 23:59:59 2026 GMT
*  subjectAltName: host "example.com" matched cert's "*.execute-api.us-west-2.amazonaws.com"          <----
*  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M03**
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://example.com/my-endpoint
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: example.com]
* [HTTP/2] [1] [:path: /my-endpoint
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
* [HTTP/2] [1] [authorization: Bearer xxxxxx]
> GET /my-endpoint
> Host: example.com
> User-Agent: curl/8.7.1
> Accept: */*
> Authorization: Bearer xxxxxx
> 
* Request completely sent off

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0< HTTP/2 200
< date: Mon, 01 Sep 2025 17:14:08 GMT
< content-type: application/json
< content-length: 42622
< access-control-allow-origin: *
< x-powered-by: PHP/8.3.19
< cache-control: no-cache, private
< apigw-requestid: xxxxxxxxxx
<
{ [3858 bytes data]
{"data":[ --> redacted
100 42622  100 42622    0     0  32083      0  0:00:01  0:00:01 --:--:-- 32094
* Connection #0 to host example.com left intact
---> redacted
```

</details>

<details>

<summary>Manually Specifying the Accept-Encoding Request - Binary Response</summary>

```
➜ curl --header "Accept-Encoding: gzip" --header "Authorization: Bearer xxxx" https://example.com/my-endpoint -vvv > r2.txt 2>&1
```

r2.txt
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host example.com:443 was resolved.
* IPv6: (none)
* IPv4: xx.xxx.xx.xx, yy.yy.yy.yy
*   Trying xx.xxx.xx.xx:443...
* Connected to example.com (xx.xxx.xx.xx) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
} [351 bytes data]
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* (304) (IN), TLS handshake, Unknown (8):
{ [19 bytes data]
* (304) (IN), TLS handshake, Certificate (11):
{ [3849 bytes data]
* (304) (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* (304) (IN), TLS handshake, Finished (20):
{ [36 bytes data]
* (304) (OUT), TLS handshake, Finished (20):
} [36 bytes data]
* SSL connection using TLSv1.3 / AEAD-AES128-GCM-SHA256 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.execute-api.us-west-2.amazonaws.com          <----
*  start date: May 20 00:00:00 2025 GMT
*  expire date: Jun 16 23:59:59 2026 GMT
*  subjectAltName: host "example.com" matched cert's "*.execute-api.us-west-2.amazonaws.com"          <----
*  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M03
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://example.com/my-endpoint
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: example.com]
* [HTTP/2] [1] [:path: /my-endpoint]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
* [HTTP/2] [1] [accept-encoding: gzip]
* [HTTP/2] [1] [authorization: Bearer xxx]
> GET /my-endpoint HTTP/2
> Host: example.com
> User-Agent: curl/8.7.1
> Accept: */*
> Accept-Encoding: gzip          <------------------
> Authorization: Bearer xxx
>
* Request completely sent off

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0< HTTP/2 200
< date: Mon, 01 Sep 2025 17:14:52 GMT
< content-type: application/json
< content-length: 11103
< access-control-allow-origin: *
< x-powered-by: PHP/8.3.19
< cache-control: no-cache, private
< content-encoding: gzip          <--------------------
< apigw-requestid: QO4l9gdOPHcESGw=
<
{ [8192 bytes data]
��	o$ɑ&�W<L
;��:=������Y���?:*�d���,����^�W@�eb6�
��$w�(�}`PK�ld+i�{������d

��l^���,9C�#/�gzD}�}�.��3�B#�֐�v

--> redacted
100 11103  100 11103    0     0  11276      0 --:--:-- --:--:-- --:--:-- 11272
* Connection #0 to host example.com left intact
�%�ˆ�}[�������@!���>�<?4��r􁕇זM��C@3,�y�$�y����&���}�Q��݌��Yx�ŧ�}؞����.=�~�

--> redacted
```
</details>

<details>

<summary>Curl with Compressed - JSON Response</summary>

```
➜ curl --compressed --header "Authorization: Bearer xxxx" https://my-example.com/my-endpoint -vvv > r3.txt 2>&1
```

r3.txt
```
➜ curl --compressed --header "Authorization: Bearer xxxx" https://my-example.com/my-endpoint -vvv > r3.txt 2>&1

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host my-example.com:443 was resolved.
* IPv6: (none)
* IPv4: xx.xxx.xx.xx, yy.yy.yy.yy8
*   Trying xx.xxx.xx.xx:443...
* Connected to my-example.com (xx.xxx.xx.xx) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
} [351 bytes data]
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* (304) (IN), TLS handshake, Unknown (8):
{ [19 bytes data]
* (304) (IN), TLS handshake, Certificate (11):
{ [3849 bytes data]
* (304) (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* (304) (IN), TLS handshake, Finished (20):
{ [36 bytes data]
* (304) (OUT), TLS handshake, Finished (20):
} [36 bytes data]
* SSL connection using TLSv1.3 / AEAD-AES128-GCM-SHA256 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.execute-api.us-west-2.amazonaws.com          <----
*  start date: May 20 00:00:00 2025 GMT
*  expire date: Jun 16 23:59:59 2026 GMT
*  subjectAltName: host "my-example.com" matched cert's "*.execute-api.us-west-2.amazonaws.com"          <----
*  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M03
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://my-example.com/my-endpoint
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: my-example.com]
* [HTTP/2] [1] [:path: /my-endpoint]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
* [HTTP/2] [1] [accept-encoding: deflate, gzip]          <----
* [HTTP/2] [1] [authorization: Bearer xxxx]
> GET /my-endpoint HTTP/2
> Host: my-example.com
> User-Agent: curl/8.7.1
> Accept: */*
> Accept-Encoding: deflate, gzip          <----
> Authorization: Bearer xxxx
>
* Request completely sent off
< HTTP/2 200
< date: Mon, 01 Sep 2025 17:14:57 GMT
< content-type: application/json
< content-length: 11103
< content-encoding: gzip          <----
< access-control-allow-origin: *
< x-powered-by: PHP/8.3.19
< cache-control: no-cache, private
< apigw-requestid: QO4mtimdPHcES1g=
<
{ [3834 bytes data]
{"data":[--> redacted
100 11103  100 11103    0     0  11826      0 --:--:-- --:--:-- --:--:-- 11824
* Connection #0 to host my-example.com left intact
--> redacted
```
</details>


Basically if curl is aware of compression it will handle it transparently, otherwise it will just obey whatever is going on. 

> Same e.g. if I use any PHP HTTP client and hit the API?

Depends on the HTTP Client and a few factors. Guzzle will decode gzipped responses by default [[source]](https://github.com/guzzle/guzzle/blob/b51ac707cfa420b7bfd4e4d5e510ba8008e822b4/docs/request-options.rst#decode_content) but will not specify it accepts gzip by default. However, as we saw above, curl is capable of handling things natively if configured right so it kind of depends whether the client will be handling things or will be delegating to curl and how that has been setup.

**Comment by mnapoli** [comment]
Yes I meant "by default" without looking into the Accept header. I think what you suggest and checking the Accept header is the better option indeed.

**Comment by victormacko** [comment]
@mnapoli Great idea RE including it natively, esp for the DX side of things.
One approach to avoid it being a breaking change would also be to check if the response header is also already set (ie. Content-Encoding = 'gzip') to not have it re-compress an already compressed response.

One thing to add - i'm not sure if API Gateway is faster (time to compress) / cheaper (ie. less processing time) gzipping things than doing it within Lambda, I didn't see a noticeable increase in processing time (ie. it all seemed about the same).

**Comment by mnapoli** [comment]
FYI I've at least documented the current behavior in https://github.com/brefphp/bref/pull/2009

I'm still considering this for v3
